### PR TITLE
Carbon Pairfinder

### PR DESCRIPTION
**Pull request by Lesigh-3100**
Automatically searches for Carbon pairs or triangles (depending on mode) and finds pools to create viable routes for them.

### CODE REVIEW

File: fastlane_bot/events/exchanges/balancer.py [hunk 1]
```diff
@@ -83,4 +83,10 @@ async def get_tkn_n(self, address: str, contract: Contract, event: Any, index: i
...
         token_balances = pool_balances[1]
-        return token_balances[index]
\ No newline at end of file
+        return token_balances[index]
+
+    def get_pool_function(self, factory_contract: Contract):
+        """
+        This function is unused for Carbon.
+        """
+        pass
```
Comment (barakman) [inline fastlane_bot/events/exchanges/balancer.py:92]: Either this function is not needed to begin with, or it is needed, and then attempting to call it and make use of the returned value will fail, since it doesn't return anything.

File: fastlane_bot/events/managers/base.py [hunk 2]
```diff
@@ -128,8 +129,10 @@ def __post_init__(self):
                 self.SUPPORTED_BASE_EXCHANGES.append(base_exchange_name)
 
             self.exchanges[exchange_name] = exchange_factory.get_exchange(key=exchange_name, cfg=self.cfg, exchange_initialized=initialize_events)
-            if base_exchange_name in SOLIDLY_V2_NAME:
+            if base_exchange_name == SOLIDLY_V2_NAME:
                 self.exchanges[exchange_name] = self.handle_solidly_exchanges(exchange=self.exchanges[exchange_name])
+            if base_exchange_name in [SOLIDLY_V2_NAME, UNISWAP_V2_NAME, UNISWAP_V3_NAME]:
+                self.exchanges[exchange_name] = self.initialize_factory_contract(exchange=self.exchanges[exchange_name])
```
Comment (barakman) [inline fastlane_bot/events/managers/base.py:135]: For `SOLIDLY_V2_NAME`, you are overriding the previous value of `self.exchanges[exchange_name]`.

So either the previous assignment is not needed, or this assignment is not needed.

File: fastlane_bot/events/exchanges/solidly_v2.py [hunk 3]
```diff
@@ -57,17 +57,70 @@ async def _get_tkn0_B(contract: Contract) -> str:
...
+
+    Args:
+        tkn0: The first token address.
+        tkn1: The second token address.
+        stable: (bool) If True, indicates a stable pool. If False, indicates a Volatile pool.
+    Returns:
+        The function returns the arguments necessary to get pool addresses from the factory contract.
+
+    """
+    return tkn0
```
Comment (barakman) [inline fastlane_bot/events/exchanges/solidly_v2.py:111]: It probably needs to be `return [tkn0]` in order for it to work when used.

File: fastlane_bot/events/exchanges/solidly_v2.py [hunk 4]
```diff
@@ -57,17 +57,70 @@ async def _get_tkn0_B(contract: Contract) -> str:
...
+    "stratum_v2"    : {"decimals":  4, "factory_abi": VELOCIMETER_V2_FACTORY_ABI, "pool_abi": SOLIDLY_V2_POOL_ABI, "get_fee": _get_fee_1, "get_tkn0": _get_tkn0_A, "get_tkn1": _get_tkn1_A, "get_pool_function": _get_pool_function_1, "get_pool_args": _get_pool_args_1},
+    "lynex_v2"      : {"decimals":  4, "factory_abi": LYNEX_V2_FACTORY_ABI      , "pool_abi": SOLIDLY_V2_POOL_ABI, "get_fee": _get_fee_5, "get_tkn0": _get_tkn0_A, "get_tkn1": _get_tkn1_A, "get_pool_function": _get_pool_function_1, "get_pool_args": _get_pool_args_1},
+    "nile_v2"       : {"decimals":  4, "factory_abi": NILE_V2_FACTORY_ABI       , "pool_abi": SOLIDLY_V2_POOL_ABI, "get_fee": _get_fee_6, "get_tkn0": _get_tkn0_A, "get_tkn1": _get_tkn1_A, "get_pool_function": _get_pool_function_1, "get_pool_args": _get_pool_args_1},
+    "xfai_v0"       : {"decimals":  4, "factory_abi": XFAI_V0_FACTORY_ABI       , "pool_abi": XFAI_V0_POOL_ABI   , "get_fee": _get_fee_7, "get_tkn0": _get_tkn0_B, "get_tkn1": _get_tkn1_B, "get_pool_function": _get_pool_function_2, "get_pool_args": _get_pool_args_2},
 }
```
Comment (barakman) [inline fastlane_bot/events/exchanges/solidly_v2.py:124]: This looks like a complete overkill.

You can probably just use the following 3 functions:
```
def _get_pool_1(factory_contract, tkn0, tkn1, stable):
    return factory_contract.functions.getPair(tkn0, tkn1, stable)

def _get_pool_2(factory_contract, tkn0, tkn1, stable):
    return factory_contract.functions.getPool(tkn0, tkn1, stable)

def _get_pool_3(factory_contract, tkn0, tkn1, stable):
    return factory_contract.functions.getPool(tkn0)
```
Then combine the `get_pool_function` attributes and the `get_pool_args` attributes into a single attribute which will be set to one of these 3 functions on each entry in the table above.

File: fastlane_bot/events/exchanges/uniswap_v2.py [hunk 5]
```diff
@@ -59,3 +58,16 @@ async def get_tkn0(address: str, contract: AsyncContract, event: Any) -> str:
...
+            This function is intended to be used with a Multicall. It fetches pools from a Uniswap V2 fork Factory contract.
+
+            Args:
+                factory_contract: The factory contract.
+
+            Returns:
+                The function.
+
+            """
+        return factory_contract.functions.getPair
```
Comment (barakman) [inline fastlane_bot/events/exchanges/uniswap_v2.py:73]: You may as well just use it directly where needed (since you're checking UniV2 before calling it).

File: fastlane_bot/events/exchanges/uniswap_v3.py [hunk 6]
```diff
@@ -55,3 +55,16 @@ async def get_tkn0(self, address: str, contract: Contract, event: Any) -> str:
...
+            This function is intended to be used with a Multicall. It fetches pools from a Uniswap V3 fork Factory contract.
+
+            Args:
+                factory_contract: The factory contract.
+
+            Returns:
+                The function.
+
+            """
+        return factory_contract.functions.getPool
```
Comment (barakman) [inline fastlane_bot/events/exchanges/uniswap_v3.py:70]: You may as well just use it directly where needed (since you're checking UniV3 before calling it).

File: fastlane_bot/pool_finder.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        pairs = [(tkn, token) for pair in unsupported_pairs for tkn in pair for token in self.flashloan_tokens]
+
+        result_list = {}
+        for ex_name, ex_data in self.multicallers.items():
+            mc = ex_data["multicaller"]
+            ex = ex_data["exchange"]
+            with mc:
+                for pair in pairs:
+                    if ex.base_exchange_name == UNISWAP_V2_NAME:
+                        mc.add_call(ex.get_pool_function(ex.sync_factory_contract), pair[0], pair[1])
```
Comment (barakman) [inline fastlane_bot/pool_finder.py:114]: You may as well just use the contract function directly.

File: fastlane_bot/pool_finder.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                    elif ex.base_exchange_name == UNISWAP_V3_NAME:
+                        for fee in self.uni_v3_fee_tiers[ex.exchange_name]:
+                            mc.add_call(ex.get_pool_function(ex.sync_factory_contract), pair[0], pair[1], fee)
```
Comment (barakman) [inline fastlane_bot/pool_finder.py:117]: You may as well just use the contract function directly.

File: fastlane_bot/pool_finder.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+                           implementation specifics of the multicall context manager or the exchange's
+                           get_pool_function method if it encounters a problem.
+            """
+        pairs = [(tkn, token) for pair in unsupported_pairs for tkn in pair for token in self.flashloan_tokens]
+
+        result_list = {}
+        for ex_name, ex_data in self.multicallers.items():
+            mc = ex_data["multicaller"]
+            ex = ex_data["exchange"]
+            with mc:
```
Comment (barakman) [inline fastlane_bot/pool_finder.py:111]: Why is this `with` statement necessary?

File: main.py [hunk 10]
```diff
@@ -706,7 +718,7 @@ def _run_async_update_with_retries(mgr, current_block, max_retries=5):
     )
     parser.add_argument(
         "--blockchain",
-        default="ethereum",
+        default="coinbase_base",
```
Comment (barakman) [inline main.py:721]: Why is this change needed?

File: main.py [hunk 11]
```diff
@@ -308,6 +310,10 @@ def run(mgr, args, tenderly_uri=None) -> None:
     start_timeout = time.time()
     mainnet_uri = mgr.cfg.w3.provider.endpoint_uri
     handle_static_pools_update(mgr)
+
+    pool_finder = PoolFinder(uni_v2_forks=mgr.cfg.network.UNI_V2_FORKS, uni_v3_forks=mgr.cfg.network.UNI_V3_FORKS, solidly_v2_forks=mgr.cfg.network.SOLIDLY_V2_FORKS, carbon_forks=mgr.cfg.network.CARBON_V1_FORKS, flashloan_tokens=args.flashloan_tokens) if args.pool_finder != -1 else None
+    if pool_finder:
```
Comment (barakman) [inline main.py:315]: Why not just do a simple if, then create and run the path finder, instead of doing an 'if', and then another 'if' based on the result of the previous 'if'?

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/events/exchanges/balancer.py (hunk 1 above)
Comment (platonfloria) [reply fastlane_bot/events/exchanges/balancer.py]: Not relevant any more

File: fastlane_bot/events/exchanges/solidly_v2.py (hunk 3 above)
Comment (platonfloria) [reply fastlane_bot/events/exchanges/solidly_v2.py]: Not relevant any more

File: fastlane_bot/events/exchanges/solidly_v2.py (hunk 4 above)
Comment (platonfloria) [reply fastlane_bot/events/exchanges/solidly_v2.py]: Not relevant any more

File: fastlane_bot/events/exchanges/uniswap_v2.py (hunk 5 above)
Comment (platonfloria) [reply fastlane_bot/events/exchanges/uniswap_v2.py]: Not relevant any more

File: fastlane_bot/pool_finder.py (hunk 8 above)
Comment (platonfloria) [reply fastlane_bot/pool_finder.py]: Not relevant any more

File: fastlane_bot/pool_finder.py (hunk 9 above)
Comment (platonfloria) [reply fastlane_bot/pool_finder.py]: Not relevant any more

File: main.py (hunk 11 above)
Comment (platonfloria) [reply main.py]: Fixed

File: fastlane_bot/pool_finder.py (hunk 7 above)
Comment (platonfloria) [reply fastlane_bot/pool_finder.py]: Not relevant any more

File: fastlane_bot/events/managers/base.py (hunk 2 above)
Comment (platonfloria) [reply fastlane_bot/events/managers/base.py]: Fixed
